File: js-json-logic-engine/jsonlogic.js
const { Before, Given, When, Then } = require('@cucumber/cucumber');
const { LogicEngine } = require('json-logic-engine');
const assert = require('assert');

class JSONLogicWorld {
    constructor() {
        this.rule = null;
        this.data = null;
        this.result = null;
    }
}

const engine = new LogicEngine(undefined, { compatible: true });

Before(function() {
    this.world = new JSONLogicWorld();
});

Given('the rule {string}', function(ruleStr) {
    const cleanStr = ruleStr.replace(/\\"/g, '"');
    try {
        this.world.rule = JSON.parse(cleanStr);
    } catch (e) {
        throw new Error(`Failed to parse rule JSON: ${e}\nInput was: ${cleanStr}`);
    }
});

Given('the data {string}', function(dataStr) {
    const cleanStr = dataStr.replace(/\\"/g, '"');
    try {
        this.world.data = JSON.parse(cleanStr);
    } catch (e) {
        throw new Error(`Failed to parse data JSON: ${e}\nInput was: ${cleanStr}`);
    }
});

When('I evaluate the rule', function() {
    const start = process.hrtime();
    this.world.result = engine.run(
        this.world.rule,
        this.world.data || null
    );
    const [seconds, nanoseconds] = process.hrtime(start);
    this.executionTime = seconds * 1e9 + nanoseconds;
});

Then('the result should be {string}', function(expected) {
    let expectedValue;
    const cleanStr = expected.replace(/\\"/g, '"');
    try {
        expectedValue = JSON.parse(cleanStr);
    } catch (e) {
        throw new Error(`Failed to parse expected data JSON: ${e}\nInput was: ${cleanStr}`);
    }

    assert.deepStrictEqual(
        this.world.result,
        expectedValue
    );
});

File: python-tests/features/steps/json-logic.py
from behave import given, when, then
from json_logic import jsonLogic
import json
import time
import os

LIBRARY = os.getenv('JSONLOGIC_LIBRARY', 'json-logic-qubit')

if LIBRARY == 'python-jsonlogic':
    from jsonlogic.operators import operator_registry
    from jsonlogic import JSONLogicExpression
    from jsonlogic.evaluation import evaluate
elif LIBRARY == 'json-logic-qubit':
    from json_logic import jsonLogic
elif LIBRARY == 'panzi-json-logic':
    from json_logic import jsonLogic

@given('the rule {rule_str}')
def given_rule(context, rule_str):
    clean_str = rule_str.strip("'")
    clean_str = clean_str.replace('\\"', '"')
    try:
        context.rule = json.loads(clean_str)
    except json.JSONDecodeError as e:
        raise Exception(f"Failed to parse rule JSON: {e}\nInput was: {clean_str}")

@given('the data {data_str}') 
def given_data(context, data_str):
    clean_str = data_str.strip("'")
    clean_str = clean_str.replace('\\"', '"')
    try:
        context.data = json.loads(clean_str)
    except json.JSONDecodeError as e:
        raise Exception(f"Failed to parse data JSON: {e}\nInput was: {clean_str}")

@when('I evaluate the rule')
def evaluate_rule(context):
    start = time.perf_counter_ns()
    print(f"Rule: {type(context.rule)}:{json.dumps(context.rule)}, Data: {type(context.data)}:{context.data}")
    try:
        if LIBRARY == 'python-jsonlogic':
            expr = JSONLogicExpression.from_json(context.rule)
            root_op = expr.as_operator_tree(operator_registry)
            context.result = evaluate(root_op, context.data, data_schema=None)
        else:
            context.result = jsonLogic(context.rule, context.data)
        context.execution_time = time.perf_counter_ns() - start
    except Exception as e:
        context.execution_time = time.perf_counter_ns() - start
        raise e

@then('the result should be {expected_str}')
def check_result(context, expected_str):
    clean_str = expected_str.strip("'")
    clean_str = clean_str.replace('\\"', '"')
    try:
        expected = json.loads(clean_str)
    except json.JSONDecodeError as e:
        raise Exception(f"Failed to parse expected JSON: {e}\nInput was: {clean_str}")
    
    assert context.result == expected